packages/eslint-plugin/src/rules/block-spacing.ts

```diff
@@ -0,0 +1,166 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+import { AST_TOKEN_TYPES } from '@typescript-eslint/utils';
+
+import * as util from '../util';
+import { getESLintCoreRule } from '../util/getESLintCoreRule';
+
+const baseRule = getESLintCoreRule('block-spacing');
+
+export type Options = util.InferOptionsTypeFromRule<typeof baseRule>;
+export type MessageIds = util.InferMessageIdsTypeFromRule<typeof baseRule>;
+
+export default util.createRule<Options, MessageIds>({
+  name: 'block-spacing',
+  meta: {
+    type: 'layout',
+    docs: {
+      description:
+        'Disallow or enforce spaces inside of blocks after opening block and before closing block',
+      recommended: false,
+      extendsBaseRule: true,
+    },
+    fixable: 'whitespace',
+    hasSuggestions: baseRule.meta.hasSuggestions,
+    schema: baseRule.meta.schema,
+    messages: baseRule.meta.messages,
+  },
+  defaultOptions: ['always'],
+
+  create(context, [whenToApplyOption]) {
+    const sourceCode = context.getSourceCode();
+    const baseRules = baseRule.create(context);
+    const always = whenToApplyOption !== 'never';
+    const messageId = always ? 'missing' : 'extra';
+    /**
+     * Gets the open brace token from a given node.
+     * @returns The token of the open brace.
+     */
+    function getOpenBrace(
+      node: TSESTree.TSEnumDeclaration,
+    ): TSESTree.PunctuatorToken {
+      // guaranteed for enums
+      // This is the only change made here from the base rule
+      return sourceCode.getFirstToken(node, {
+        filter: token =>
+          token.type === AST_TOKEN_TYPES.Punctuator && token.value == '{',
+      }) as TSESTree.PunctuatorToken;
+    }
+
+    /**
+     * Checks whether or not:
+     *   - given tokens are on same line.
+     *   - there is/isn't a space between given tokens.
+     * @param left A token to check.
+     * @param right The token which is next to `left`.
+     * @returns
+     *    When the option is `"always"`, `true` if there are one or more spaces between given tokens.
+     *    When the option is `"never"`, `true` if there are not any spaces between given tokens.
+     *    If given tokens are not on same line, it's always `true`.
+     */
+    function isValid(left: TSESTree.Token, right: TSESTree.Token): boolean {
+      return (
+        !util.isTokenOnSameLine(left, right) ||
+        sourceCode.isSpaceBetween!(left, right) === always
+      );
+    }
+
+    /**
+     * Checks and reports invalid spacing style inside braces.
+     * @param {ASTNode} node A BlockStatement/StaticBlock/SwitchStatement node to check.
+     * @returns
```

Comment:
Nit: we have types in TypeScript. `ASTNode` is imprecise and the `@returns` is implied.

```suggestion
     * @param node A BlockStatement/StaticBlock/SwitchStatement node to check.
```

Seems like these were copy & pasted from https://github.com/eslint/eslint/blob/ba74253e8bd63e9e163bbee0540031be77e39253/lib/rules/block-spacing.js#L86? I don't see other base rule files giving explicit attribution, so I guess this should be fine as-is.

Reply:
I saw other rules that copied from the base rule, so I assumed it's ok, do you suggest only removing the types from the jsdoc?

Reply:
Yeah there's no need for the jsdoc.

Reply:
Ok just following up from internal discussions! Still no need to take action here. Just posting as a reference...

Some of our extension rules include code that's very similar to the code from the base ESLint rule. Sometimes there are entire lines or even functions that look the same.

[ESLint is licensed under the MIT License](https://github.com/eslint/eslint/blob/b401cde47d44746ff91b8feced3fb3a4e32c0e12/LICENSE). MIT stipulates _(emphasis mine)_:

> Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:
> 
> The above copyright notice and this permission notice shall be included in
all **copies or substantial portions of the Software**.

We're not legal experts -let alone copyright lawyers- but we believe we're not in violation of this because:

* Our code does not copy substantial portions. At most, the _ideas_ of the original code are allowed in as reference, but we use our own code style & standards in the implementation.
* We link to the original ESLint source quite often in our site & packages, including in the individual docs pages for rules.

_If anybody knows my statements to be wrong and/or that we need to do something else, please let us know!_